rfcs/0000-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,75 @@
+3. The maintainer will get the PR reviewed and approved.
+    1. It is **recommended but not required** to have some maintainers of dependent packages involved in the review to review if the breakage is justified.
+4. The maintainer will contact the maintainers of all dependent, broken packages (herein called sub-maintainers).
+    1. The sub-maintainers will be provided a link to the PR with the breaking changes as well as any context that will help them resolve the breakages.
+    2. The sub-maintainers are expected to update the derivations that they maintain within 7 days.
+        1. These changes should be merged to `master` following the regular nixpkgs PR process if they are backwards compatible. Otherwise they will be merged into the branch of the breaking PR.
+5. The maintainer must provide a grace period of at least 7 days from when the sub-maintainers were notified.
+6. The maintainer must merge `master` into their PR branch.
+7. The maintainer must mark all still-broken packages as [broken](https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes).
+8. The maintainer can now merge to `master`.
+
+This procedure should not result in a failing package build in `master` at any point.
+
+# Drawbacks
+
+This delays the merge of “core” derivations as the author needs to wait for sub-packages to be tested and possibly fixed up to the 7 day threshold.
+
+# Alternatives
+## Shorter grace period
+
+A shorter grace period allows maintainers to move more quickly but requires maintainers of defendant derivations to jump to action quickly which is not always possible for volunteer driven work. For example if a maintainer is on vacation it can’t be expected that they respond in a couple of days.
+
+## Longer grace period
+
+A longer grace period would give derivation maintainers more time to react to changes in dependencies and provide more time to search for replacement maintainers if the original maintainer has abandoned the derivation. The 7 day number was mostly arbitrary and can easily be changed in the future. A longer grace period would also result in more packages being fixed before the breaking change is shipped which would mean that end-users would experience less breakage on channels.
+
+# Unresolved questions
+## Critical Packages and Tests
+
+What if a breaking change breaks NixOS tests? There must be packages and tests so critical that we can not merge without them passing? In that case do we leave the PR open until fixed? or use the `staging` branch?
+
+# Future work
+- Create a tool for automatically notifying maintainers of broken dependents
+- Create a tool for automatically marking broken packages as broken.
+- [This is a milestone on the way to implemented a queue-based merge process for nixpkgs.](https://paper.dropbox.com/doc/MTSY8xKH6y1xDEwavyDNW)
```

Comment:
I have the fear that this link will bit rot over time.

Reply:
Do you have a preferred link? I could just leave it as a description but I figured the link didn't harm.

Reply:
I have mirrored it to web.archive.org https://web.archive.org/web/20210325015950if_/https://paper.dropbox.com/doc/nixpkgs-Merge-Queue-Overview-MTSY8xKH6y1xDEwavyDNW